Sources/CvPlayerAI.cpp

```diff
@@ -12852,9 +12852,58 @@ int CvPlayerAI::AI_unitPropertyValue(UnitTypes eUnit, PropertyTypes eProperty) c
 	return iValue;
 }
 
-/********************************************************************************/
-/* 	City Defenders						24.07.2010				Fuyu			*/
-/********************************************************************************/
+int CvPlayerAI::AI_unitBuildingValue(UnitTypes eUnit, CvArea* pArea) const
+{
+	FAssertMsg(eUnit != NO_UNIT, "Unit is not assigned a valid value");
+	FAssertMsg(eUnitAI != NO_UNITAI, "UnitAI is not assigned a valid value");
+
+	CvUnitInfo& kUnitInfo = GC.getUnitInfo(eUnit);
+
+	// Evaluate unit's ability to construct buildings
+	int iCount = 0;
+
+	if (kUnitInfo.getNumBuildings() > 0 && pArea != NULL && !isNPC())
+	{
+		int iBuildingValue;
+		int iLoop;
+		bool bCoastal = kUnitInfo.getDomainType() == DOMAIN_SEA;
+		int iMinOceanSize;
+		if (bCoastal)
+		{
+			iMinOceanSize = GC.getMIN_WATER_SIZE_FOR_OCEAN();
+		}
+
+		for (int iI = 0; iI < kUnitInfo.getNumBuildings(); iI++)
+		{
+			BuildingTypes eBuilding = (BuildingTypes)kUnitInfo.getBuildings(iI);
+
+			if (NO_BUILDING != eBuilding)
+			{
+				if (canConstruct(eBuilding, false, false, true) && AI_getNumBuildingsNeeded(eBuilding, bCoastal) > 0)
+				{
+					iCount = 0;
```

Comment:
Why are we resetting iCount here? It means we are only counting the number of places we can build whatever happens to be last in the units building list. This makes no sense!